### PR TITLE
feat: add Validating/Mutating webhook analyzer

### LINKED ISSUE #533: [Feature]: Mutating/Validating web hook analyser
### Checklist

- [X] I've searched for similar issues and couldn't find anything matching
- [X] I've discussed this feature request in the [K8sGPT Slack](https://k8sgpt.slack.com) and got positive feedback

### Is this feature request related to a problem?

No

### Problem Description

_No response_

### Solution Description

It is possible for the MutatingAdmission and Validating admission web hooks to point to receiver pods that are not active.
It would be possible to create a new analyser type to detect issues

### Benefits

This would identify issues found with dead hooks

### Potential Drawbacks

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by rakshitgondwal**
Closes #533 

## 📑 Description
Changes made:
- Added an analyzer for Validating Webhook that checks if the pod that webhook points to is not active.
- Added an analyzer for Mutating Webhook that checks if the pod that webhook points to is not active.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

## Additional Information
Screenshot from 2023-07-07 17-47-44

### CODE REVIEW

**Review by AlexsJones — APPROVED** [review]

**Comment by arbreezy** [comment]
@rakshitgondwal would be nice to share what openai's response is as well based on the error messages you feed it

**Comment by rakshitgondwal** [comment]
> @rakshitgondwal would be nice to share what openai's response is as well based on the error messages you feed it

Screenshot from 2023-07-10 22-15-31

**Comment by rakshitgondwal** [comment]
> I also left a few comments, thanks for your contribution !

Sorry, but I can't see your comments, can you please re check?

**Review by arbreezy — CHANGES_REQUESTED** [review]

File: pkg/analyzer/validating-webhook.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+	for key, value := range preAnalysis {
+		var currentAnalysis = common.Result{
+			Kind:  kind,
+			Name:  key,
+			Error: value.FailureDetails,
+		}
+
+		parent, _ := util.GetParent(a.Client, value.ValidatingWebhook.ObjectMeta)
```
[the remark this answers is not in the document]
Comment (rakshitgondwal) [reply pkg/analyzer/validating-webhook.go:100]: Can you help me out here? I didn't get this.

[the remark this answers is not in the document]
Comment (AlexsJones) [reply pkg/analyzer/validating-webhook.go:104]:
```
func GetParent(client *kubernetes.Client, meta metav1.ObjectMeta) (string, bool) {
	if meta.OwnerReferences != nil {
		for _, owner := range meta.OwnerReferences {
			switch owner.Kind {
			case "ReplicaSet":
```

You need to add a case in for these types he means @rakshitgondwal

[the remark this answers is not in the document]
Comment (rakshitgondwal) [reply pkg/analyzer/validating-webhook.go:104]: Thank you. Pushed up a commit fixing this. Please have a look.

**Review by AlexsJones — APPROVED** [review]

**Review by arbreezy — APPROVED** [review]

**Comment by rakshitgondwal** [comment]
Thank you for the merge!

**Comment by AlexsJones** [comment]
image


As per here, use the service-name the webhook is bound too and check if the service has pods that are active ( at least 1 )
that should be the real logic

**Comment by rakshitgondwal** [comment]
I was trying out this logic, but it is not working. I added a `LabelSelector` while getting the pods to be `service.name`, then looped on the pods and check if they are running. Can you help me out with this? @AlexsJones 
```
svc := webhook.ClientConfig.Service
pods, err := a.Client.GetClient().CoreV1().Pods(a.Namespace).List(context.Background(), v1.ListOptions{LabelSelector: svc.Name})
	if err != nil {
		return nil, err
	}
	activePods := 0
	for _, pod := range pods.Items {
		if pod.Status.Phase == "Running" {
			activePods++
		}
	}
	if activePods == 0 {
		doc := apiDoc.GetApiDocV2("spec.webhook")
		failures = append(failures, common.Failure{
```

**Comment by AlexsJones** [comment]
I can take a look at implementing it a version of this and seeing whats wrong later today

**Pull request merged**